keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,982 @@
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+Using kubectl's normal error handling. There is no lasting effect to data or the
+user.
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+N/A
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Implement proto.Models for OpenAPI V3 data
+
+The current hard-coded printer is capable of printing any objects in `proto.Models` form.
+
+[We already have a way to express OpenAPI v3 data as `proto.Models`, so this can be
+seen as a path of least resistance for plugging OpenAPI v3 into `kubectl explain`.
+
+This approach is undesirable for a few different reasons:
+
+1.) We would like to update the explain printer to include new OpenAPI v3 information,
+the current design makes that time consuming and not maintainable.
+
+2.) API-Machinery has desire to deprecate `proto.Models`. We see`proto.Models`
+conversion as unnecessary and costly buraucracy, that contributes to high
+OpenAPI overhead. We are seeking to deprecate the type in favor of the
+kube-openapi types for future usage.
+
+### Custom User Templates
+Users might also like to be able to specify a path to a custom template file for
+the resource information to be written to:
+
+human-readable plaintext form:
+```shell
+kubectl explain pods --template /path/to/template.tmpl
```

Comment:
After the recent change to the planned flag and having reflected more on the html/md options, I wanted to revisit this to make sure the flag set we've chosen isn't going to paint us into a corner. Even if we do add a default markdown implementation, it's easy to imagine that some users will prefer things to be displayed differently than we've chosen, leading to demand for this feature. 

So, the new thought that has occurred to me is that it will technically be necessary for the user to be able to tell us which version of openAPI their template expects, similar to the problem we had with the long-term strategy around "raw". What would that look like? `--output=openapiv3 --template /path/to/template.tmpl` doesn't quite feel right... Our other commands that support file-based templating use `--output=go-template-file --template=/path/to/template`, so I think we'd want that for consistency. But it isn't sufficient, since as I was saying, the schema version isn't already inherent in the command, unlike in every case that currently exists.

So, maybe we need (at least) two flags after all, but not the ones we had this morning? I feel like I've talked myself in a circle. 😅 
- `--output=plaintext|openapi-json|md|html|go-template-file` --> during alpha and beta, requires a compatibility check with the openapi version flag; "plaintext" would be compatible with v2 or v3, all others require v3. But the next migration to the theoretical v4 wouldn't require that, since all templates would be required to support it.
- `--openapi-version=v2|v3`--> This could be used to provide an opt-out via a "v2" value during the beta rollout period. At GA it becomes useless, so could be removed after deprecation (then re-introduced if we eventually need it again).
- And if we do add templating: `--template=/path/to/template.tmpl`

(all would presumably need the experimental prefix during alpha)

Reply:
I think our built-in templates can hardcode their supported openapi versions. If in the future we add custom templates and `--template`, then I think it would be appropriate to think about how associated metadata is included, and possibly then add the `--openapi-version` flag. I think the current proposal does not preclude that possibility.

Reply:
We already have the following in kubectl:
```
  -o, --output='': Output format. One of:
json|yaml|wide|name|custom-columns=...|custom-columns-file=...|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=...
See custom columns [http://kubernetes.io/docs/user-guide/kubectl-overview/#custom-columns], golang template
[http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template
[http://kubernetes.io/docs/user-guide/jsonpath].
```

How about:
```
kubectl explain pods -o v3-template=/path/to/template.tmpl
```

Reply:
> What would it look like to hard code supported API versions? Do you mean in some convention surrounding how kubectl accesses them (not viable for custom templates), or within the template proper?

It could be a map in the Go code that uses them. Or a naming convention from filename. Or a yaml that lives next to the templates with metadata. I am not attached to a specified idea. But the idea is that it is internal and does not need to extend to custom templates, as long as it an implementation change in the future to allow custom templates.

> How about:
>
>     kubectl explain pods -o v3-template=/path/to/template.tmpl

If we were ever to add templates in the future, I like this option since it is consistent with the existing `-o` flag as you pointed out.

Reply:
@KnVerey does your initial concern about leaving door open for future version-pinned templates still stand after hearing the above? Personally I am satisfied that we have not backed ourselves into a corner in terms of implementation or interface on this issue

Reply:
Ah, you're right: I missed that the filepath could also be specified inline. I'm not sold on the specific option name because v3 refers to an openapi version, and openapi isn't anywhere in the command, so it sounds like it might instead refer to v3 of a templating system, or maybe of pod. Although part of me feels like we're mashing two concepts into one option, the main practical drawback of fully specifying `--output=openapiv3-go-template=` is verbosity. In any case, I won't block on this. The KEP as it stands now is simpler for the features certainly planned. Plus we have an opportunity to revisit naming if need be when we promote the flag out of experimental, at which point we should have a more solid stance on the custom template possibility.

Reply:
Thanks, I wasn't trying to find the best name, but I think the mechanism leaves the door open for the future. Thanks!